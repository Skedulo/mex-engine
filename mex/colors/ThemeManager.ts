import {Appearance, ColorSchemeName} from "react-native";
import LightColors from "./LightColors";
import DarkColors from "./DarkColors";
import StylesManager from "../StylesManager";
import {SkedColorSet} from "./SkedColorSet";
import { IThemeManager } from "@skedulo/mex-engine-proxy";
class ThemeManager implements IThemeManager {

    _colorSet: SkedColorSet = LightColors
    _callbacks: ((colorScheme:ColorSchemeName) => void)[] = []

    constructor() {
    }

    initializeColors() {
        this._colorSet = Appearance.getColorScheme() === 'light' ? LightColors : DarkColors;
        StylesManager.initializeStyles();

        Appearance.addChangeListener(preferences => {
            this._colorSet = preferences.colorScheme === 'light' ? LightColors : DarkColors;
            StylesManager.initializeStyles();

            this.triggerThemeChange(preferences.colorScheme)
        })
    }

    registerThemeChangedCallback(callback:(colorScheme:ColorSchemeName) => void) {
        this._callbacks.push(callback)
    }

    removeThemeChangedCallBack(callback:(colorScheme:ColorSchemeName) => void) {
        let index = this._callbacks.indexOf(callback)

        if (index !== -1) {
            this._callbacks.slice(index, 1)
        }
    }

    triggerThemeChange(colorScheme: ColorSchemeName) {
        this._callbacks.forEach(c => c(colorScheme))
    }

    getTheme(): ColorSchemeName {
        return Appearance.getColorScheme()
    }

    getColorSet(): SkedColorSet {
        return this._colorSet
    }
}


export default new ThemeManager() as ThemeManager
