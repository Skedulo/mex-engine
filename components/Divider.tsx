import * as React from "react";
import {ColorValue, View, ViewStyle} from "react-native";
import ThemeManager from "../mex/colors/ThemeManager";

type Props = {
    color?: ColorValue
    style?: ViewStyle
} | undefined


const Divider: React.FC<Props> = (props: Props = undefined) => {

    let colors = ThemeManager.getColorSet();

    const defaultProps:Props = {
        color: colors.navy75,
        style: {}
    }

    props = { ...defaultProps, ...(props ?? {}) }

    return <View style={{
        width: "100%",
        height: 1,
        backgroundColor: props.color,
        ...props.style}}></View>
}

export default Divider
