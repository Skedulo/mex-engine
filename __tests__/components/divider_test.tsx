import React from 'react';
import {render} from '@testing-library/react-native';
import Divider from "../../components/Divider";
import ThemeManager from "../../mex/colors/ThemeManager";

describe('Divider component tests', function () {

    it('Divider component basic renderer', async function () {
        const tree = render(<Divider />)

        expect(tree.toJSON()).toMatchSnapshot()
    });

    it('Divider component with different color renderer', async function () {
        let colors = ThemeManager.getColorSet()

        const tree = render(<Divider color={colors.navy600} />)

        expect(tree.toJSON()).toMatchSnapshot()
    });
});
