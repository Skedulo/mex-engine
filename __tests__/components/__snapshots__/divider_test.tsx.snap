// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Divider component tests Divider component basic renderer 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#E5E7EE",
      "height": 1,
      "width": "100%",
    }
  }
/>
`;

exports[`Divider component tests Divider component with different color renderer 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#5E687A",
      "height": 1,
      "width": "100%",
    }
  }
/>
`;
